src/instruction.rs

```diff
@@ -301,6 +304,17 @@ impl fmt::Display for MemoryReference {
     }
 }
 
+impl FromStr for MemoryReference {
+    type Err = SyntaxError<Self>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let tokens = lex(s)?;
+        let (_, memory_reference) = all_consuming(parse_memory_reference_with_brackets)(&tokens)
+            .map_err(extract_nom_err)?;
+        Ok(memory_reference)
```

Comment:
Need to double check, but I think you also want a `disallow_leftover()` here, maybe _instead_ of `all_consuming`.

It's the same effect, but `disallow_leftover` gives a better error and removes the need for `map_err`.

```suggestion
        disallow_leftover(parse_memory_reference_with_brackets(&tokens))
```

Reply:
To make it work in this context, you'll need to change `disalllow_leftover` to be this. (Alternatively, change the error type to `ProgramError<T>`, but this is the more correct option, IMO)

```rust
pub fn disallow_leftover<I, O, E>(result: nom::IResult<I, O, ParseError>) -> Result<O, E>
where
    I: ErrorInput,
    E: From<SyntaxError<O>>,
{
    match result.finish() {
        Ok((leftover, parsed)) => {
            if leftover.is_empty() {
                Ok(parsed)
            } else {
                Err(E::from(SyntaxError::from(LeftoverError::new(leftover, parsed))))
            }
        }
        Err(err) => Err(E::from(SyntaxError::from(err))),
    }
}
```